### PR TITLE
Fixed helper `Paginator::numbers()` to get it working with a string number

### PR DESCRIPTION
**Pull request by Xety**
Also, added testscase for `Paginator::first()` & `Paginator::last()` for the same thing.

Ref #7842

### CODE REVIEW

File: src/View/Helper/PaginatorHelper.php [hunk 1]
```diff
@@ -820,9 +822,10 @@ protected function _firstNumber($ellipsis, $params, $start, $options)
     protected function _lastNumber($ellipsis, $params, $end, $options)
     {
         $out = '';
+        $last = is_int($options['last']) ? $options['last'] : 0;
```
Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: Should this be is_numeric() to count in `'1'` kind of ints, as well?

Comment (Xety) [inline src/View/Helper/PaginatorHelper.php]: No, with `is_numeric()` it does the same as before :
```php
var_dump(is_numeric('zz')); // false
var_dump(is_numeric('1')); // true
var_dump(is_numeric(1));  // true

var_dump(is_int('zz')); // false
var_dump(is_int('1')); // false
var_dump(is_numeric(1));  // true
```


The expected behavior, is to detect `'1'`, `'2'`, ... ` '9'` as a string and not as an integer.

Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: So you are saying that inputting `1` vs `'1'` will lead to two different kind of behavior?
With PHP and not being typesafe this sounds quite a dangerous thing to do then. Or maybe I don't quite understand the issue here.

Comment (Xety) [inline src/View/Helper/PaginatorHelper.php]: Yes, `1` & `'1'` will lead to 2 different kind of behavior.

It's needed if we want to use a custom template with a `'1'` - `'9'` as a value with the function `Paginator::numbers()` :
**Before the fix :**
```php
<?= $this->Paginator->numbers([
    'modulus' => 2,
    'first' => '1',
    'last' => (string)$this->Paginator->counter('{{pages}}'),
    'templates' => [
        'first' => '<li class="customFirst"><a href="{{url}}">{{text}}</a></li>',
        'last' => '<li class="customLast"><a href="{{url}}">{{text}}</a></li>'
    ]
]); ?>

// Will use the `first` template, but NOT the `last` template.
```
Result :
534c5581afc60b2e5605fb4eb640bfff


**After the fix :**
```php
<?= $this->Paginator->numbers([
    'modulus' => 2,
    'first' => '1',
    'last' => (string)$this->Paginator->counter('{{pages}}'),
    'templates' => [
        'first' => '<li class="customFirst"><a href="{{url}}">{{text}}</a></li>',
        'last' => '<li class="customLast"><a href="{{url}}">{{text}}</a></li>'
    ]
]); ?>

// The `first` and `last` templates are used correctly.
```

Result :
32ab540650126711e07a124e4d9b8cc4

Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: Do we have to backport this to 2.x, as well, afterwards?

Comment (Xety) [inline src/View/Helper/PaginatorHelper.php]: I think so.
I just tested on an old Cake2 app (V 2.4.0), and it seems that there's is also a bug related to that, but a bit different.

#### With an integer :
```php
<?= $this->Paginator->numbers(array(
    'modulus' => 2,
    'tag' => 'li',
    'currentTag' => 'a',
    'separator' => '',
    'ellipsis' => '<li><a>...</a></li>',
    'first' => 1, //With an integer
    'last' => 1 // With an integer
));?>
```
Result : It look good, the `ellipsis` is added correctly.
9a5201daae91dd515f4201147ef207c7

#### With a string number :
```php
<?= $this->Paginator->numbers(array(
    'modulus' => 2,
    'tag' => 'li',
    'currentTag' => 'a',
    'separator' => '',
    'ellipsis' => '<li><a>...</a></li>',
    'first' => '1', // With a string number
    'last' => '11' // With a string number
));?>
```
Result : The `ellipsis` are not rendered, **and** it display all the page after the current page
bbd623529159615792c724d4005082a8


#### With a *real* string :
```php
<?= $this->Paginator->numbers(array(
    'modulus' => 2,
    'tag' => 'li',
    'currentTag' => 'a',
    'separator' => '',
    'ellipsis' => '<li><a>...</a></li>',
    'first' => 'abc', // With a 'real' string
    'last' => 'abc' // With a 'real' string
));?>
```

Result : The `ellipsis` are not rendered, **but** it doesn't display all the page after the current page.
d3a433ac79a73918d9d31427da9b3179

So for me yes, there's a bug. The `1` & `'1'` seems to be identically the same by Cake2, probably by this condition : https://github.com/cakephp/cakephp/blob/2.7/lib/Cake/View/Helper/PaginatorHelper.php#L801

Also, the `ellipsis` are not rendered with a string. I don't know if it's intended or not. But if it's not intended, then it's another bug.

**Comment by markstory** [comment]
Looks decent to me.

**Pull request merged**